File: Loadtests/PlayLoad/app/models/Monitor.java
package models;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.LinkedBlockingQueue;

public class Monitor {

	private static Monitor _instance;
	private Object sentCMut = new Object(), sentSeMut = new Object(), recMut = new Object();
	public long connected = 0;
	public long clientStartTime;
	public long serverStartTime;
	public int numberOfClients;
	public BlockingQueue<TestDataEntity> testDataEntities;
	public int spacing;
	public ConcurrentHashMap<Integer, Integer> sentFromServerEvents;
	public ConcurrentHashMap<Integer, Integer> sentFromClientEvents;
	public ConcurrentHashMap<Integer, Integer> receivedAtServerEvents;
	public int harvested;
	public long duration;
	public BlockingQueue<String> completedClients;

	public static Monitor getInstance() {
		if(_instance == null) {
			_instance = new Monitor();
		}
		
		return _instance;
	}
	
	private Monitor() {
		reset();
	}

	public void reset() {
		numberOfClients = 0;
		testDataEntities = new LinkedBlockingQueue<TestDataEntity>();
		spacing = 0;
		sentFromServerEvents = new ConcurrentHashMap<Integer, Integer>();
		sentFromClientEvents = new ConcurrentHashMap<Integer, Integer>();
		receivedAtServerEvents = new ConcurrentHashMap<Integer, Integer>();
		harvested = 0;
		duration = 0;
		completedClients = new LinkedBlockingQueue<>();		
	}

	public int registerSentFromClientEvent(long millisecondsSinceEpoch) {
		return registerSentFromClientEvent(millisecondsSinceEpoch, 1);		
	}
	
	public int registerSentFromClientEvent(long millisecondsSinceEpoch, int spacing) {
		int key = getKey(millisecondsSinceEpoch, spacing, true);
		synchronized (sentCMut) {
			addEvent(sentFromClientEvents, key);
		}
		
		return key;
	}
	
	public void registerReceivedAtServerEvent(long millisecondsSinceEpoch) {
		registerReceivedAtServerEvent(millisecondsSinceEpoch, 1);
	}

	public void registerReceivedAtServerEvent(long millisecondsSinceEpoch, int spacing) {
		int key = getKey(millisecondsSinceEpoch, spacing, false);
		synchronized (recMut) {
			addEvent(receivedAtServerEvents, key);	
		}
	}

	public void registerSentFromServerEvent(long millisecondsSinceEpoch, boolean broadcast) {
		registerSentFromServerEvent(millisecondsSinceEpoch, broadcast, 1);
	}
	
	public void registerSentFromServerEvent(long millisecondsSinceEpoch, boolean broadcast, int spacing) {
		int key = getKey(millisecondsSinceEpoch, spacing, false);
		int nrOfEvents = broadcast ? numberOfClients : 1;
		synchronized (sentSeMut) {
			addEvent(sentFromServerEvents, key, nrOfEvents);	
		}
	}
	
	public void addEvent(ConcurrentHashMap<Integer, Integer> eventStore, int key) {
		addEvent(eventStore, key, 1);		
	}

	public void addEvent(ConcurrentHashMap<Integer, Integer> eventStore, int key, int nrOfEvents) {
		while(key > eventStore.values().size()) {
			eventStore.put(eventStore.values().size(), 0);			
		}
		if(eventStore.values().size() == key) {
			eventStore.put(key, nrOfEvents);
		} else {
			int value = eventStore.get(key);
			value += nrOfEvents;
			eventStore.replace(key, value);
		}
		
	}
	
	private int getKey(long millisecondsSinceEpoch, int spacing, boolean client) {
		long start = client ? clientStartTime : serverStartTime;
		long millisecondsSinceStart = millisecondsSinceEpoch - start;
		int seconds = round(false, (double)(millisecondsSinceStart/1000));
		return round(false, (double)(seconds/spacing));
	}
	
	private static int round(boolean up, double value) {
		return up ? (int)Math.ceil(value) : (int)Math.floor(value);
	}

	public boolean harvestedAll() {
		return harvested == numberOfClients;
	}
	
	public boolean complete() {
		return completedClients.size() == numberOfClients;
	}

}

File: Loadtests/SockJs/Server/monitor.js
var Monitor = function() {
    var self = this;
    self.numberOfClients = 0;
    self.completedClients = [];
    self.duration = 0;
    self.testDataEntities = [];
    self.clientStartTime = 0;
    self.serverStartTime = 0;
    self.spacing = 0;
    self.harvested = 0;
    self.sentFromClientEvents = [];
    self.receivedAtServerEvents = [];
    self.sentFromServerEvents = [];
};

Monitor.prototype.reset = function() {
    var self = this;
    self.numberOfClients = 0;
    self.completedClients = [];
    self.duration = 0;
    self.testDataEntities = [];
    self.spacing = 0;
    self.harvested = 0;
    self.sentFromClientEvents = [];
    self.receivedAtServerEvents = [];
    self.sentFromServerEvents = [];
}

Monitor.prototype.registerSentFromClientEvent = function(millisecondsSinceEpoch, spacing) {
    var sp = spacing ? spacing : 1;
    var key = getKey(millisecondsSinceEpoch, sp, this, true);
    this.addEvent(this.sentFromClientEvents, key);
    return key;
}

Monitor.prototype.registerReceivedAtServerEvent = function(millisecondsSinceEpoch, spacing) {
    var sp = spacing ? spacing : 1;
    var key = getKey(millisecondsSinceEpoch, sp, this, false);
    this.addEvent(this.receivedAtServerEvents, key);
}

Monitor.prototype.registerSentFromServerEvent = function(millisecondsSinceEpoch, broadCast, spacing) {
    var sp = spacing ? spacing : 1;
    var key = getKey(millisecondsSinceEpoch, sp, this, false);
    var nrOfEvents = broadCast ? this.numberOfClients : 1;
    this.addEvent(this.sentFromServerEvents, key, nrOfEvents);
}

Monitor.prototype.addEvent = function (eventStore, key, numberOfEvents) {
    var nrOE = numberOfEvents ? numberOfEvents : 1;

    while(key > eventStore.length) {
        eventStore.push(0);
    }
    if(eventStore.length === key) {
        eventStore.push(nrOE);
    } else {
        eventStore[key] += nrOE;
    }
}

Monitor.prototype.complete = function() {
    return this.completedClients.length === this.numberOfClients;
}

Monitor.prototype.harvestedAll = function() {
    return this.harvested === this.numberOfClients;
}

function getKey(millisecondsSinceEpoch, spacing, monitor, client) {
    var start = client ? monitor.clientStartTime : monitor.serverStartTime;
    var millisecondsSinceStart = millisecondsSinceEpoch - start;
    var seconds = round(false, millisecondsSinceStart / 1000);
    return round(false, seconds / spacing);
}

function round(up, value) {
    return up ? Math.ceil(value) : Math.floor(value);
}

var instance = null;

exports.getInstance = function() {
    if(instance) {
        return instance;
    }
    instance = new Monitor();
    return instance;
}

